### PR TITLE
[WIP] Railroads

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
### Summary:

```SUMMARY: Features "Railroad infrastructure (railroads, bridges, crossings, stations)"```

# Description:

Resurrection of #6925. Not only overmap tiles (*as in mentioned PR*), but also some basic railroad mapgen.

Still need a lot of work as this branch is 6 month old and a lot of things has been changed in mapgen since that time.

# Preview download:

You can take compiled and rather old preview version for Windows (64 bit) [here](https://www.dropbox.com/s/0t7g95nslx2rfvp/dda-railroad-test-windows-x64.zip?dl=0).

Not as good as in OpenTTD yet, but I hope it looks promising.

# To do:

- [x] - basic overmap railroad tiles;
- [x] - basic railroad mapgen;
- [x] - merge changes from current master;
- [x] - add railroad bridges;
- [x] - add railroad junctions;
- [x] - add road-rail crossing/intersections;
- [ ] - add railroad stations;
- [x] - rework railroad overmap functions (make railroad lines more straight);
- [ ] - rework railroad overmap functions (rework station placement);
- [x] - rework railroad overmap functions (disallow railroads override parallel roads);
- [ ] - rework railroad intersection placement (allow only on ns- and ew-roads);
- [ ] - rework railroad mapgen functions (remove hard-codes, cleanup, etc).

# Screenshots:

- __NS-rail tracks + minimap:__
image

- __WE-rail tracks + minimap:__
image

- __Overmap version 0__ (`light green` lines - railroad tracks):
image

### CODE REVIEW

File: data/json/terrain.json [hunk 1]
```diff
@@ -6413,5 +6413,165 @@
...
+        "move_cost" : 2,
+        "flags" : ["TRANSPARENT","FLAT"]
+    },{
+        "type" : "terrain",
+        "id" : "t_railroad_tie_d2",
+        "name" : "railroad tie",
+        "symbol" : "\\",
+        "color" : "dgray_brown",
+        "move_cost" : 2,
+        "flags" : ["TRANSPARENT","FLAT"]
```
Comment (Coolthulhu) [inline data/json/terrain.json]: Many new terrain types that only differ by color/name, are all indestructible (ie. no bash data), functionally totally equivalent to each other.
Does terrain data not support `copy-from`? Might be worth implementing it.

File: src/mapdata.cpp [hunk 2]
```diff
@@ -477,7 +477,11 @@ ter_id t_null,
     // Temple tiles
     t_rock_red, t_rock_green, t_rock_blue, t_floor_red, t_floor_green, t_floor_blue,
     t_switch_rg, t_switch_gb, t_switch_rb, t_switch_even, t_open_air, t_plut_generator,
-    t_pavement_bg_dp, t_pavement_y_bg_dp, t_sidewalk_bg_dp, t_guardrail_bg_dp;
+    t_pavement_bg_dp, t_pavement_y_bg_dp, t_sidewalk_bg_dp, t_guardrail_bg_dp,
+    // Railroad
+	t_rubble,
+    t_railroad_track, t_railroad_track_on_tie, t_buffer_stop_h, t_buffer_stop_v, t_railroad_tie, t_railroad_tie_h, t_railroad_tie_v, t_railroad_tie_d1, t_railroad_tie_d2, t_railroad_crossing_signal, t_crossbuck_wood, t_crossbuck_metal
```
Comment (Coolthulhu) [inline src/mapdata.cpp]: Needing to define those is a big red flag that it isn't all done as clearly as it could and should be.

File: src/mapdata.cpp [hunk 3]
```diff
@@ -739,6 +743,19 @@ void set_ter_ids() {
...
+    t_buffer_stop_h = ter_id( "t_buffer_stop_h" );
+    t_buffer_stop_v = ter_id( "t_buffer_stop_v" );
+    t_railroad_tie = ter_id( "t_railroad_tie" );
+    t_railroad_tie_h = ter_id( "t_railroad_tie_h" );
+    t_railroad_tie_v = ter_id( "t_railroad_tie_v" );
+    t_railroad_tie_d1 = ter_id( "t_railroad_tie_d1" );
+    t_railroad_tie_d2 = ter_id( "t_railroad_tie_d2" );
+    t_railroad_crossing_signal = ter_id( "t_railroad_crossing_signal" );
+    t_crossbuck_metal = ter_id( "t_crossbuck_metal" );
+    t_crossbuck_wood = ter_id( "t_crossbuck_wood" );
```
Comment (Coolthulhu) [inline src/mapdata.cpp]: Same here.

File: data/json/terrain.json (hunk 1 above)
Comment (ZhilkinSerg) [reply data/json/terrain.json]: I will note that and check what can be done after I rebase to current master.

File: src/mapgen_functions.cpp [hunk 4]
```diff
@@ -5381,3 +5390,335 @@ void madd_field( map *m, int x, int y, field_id t, int density )
...
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_rubble );
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_railroad_track );
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_rubble );
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_rubble );
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_railroad_track );
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_rubble );
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_rubble );
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_dirt );
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_dirt );
+                    xn++; coord_rotate_cw( xn, yn, dir ); m->ter_set( xn, yn, t_dirt );
```
Comment (Coolthulhu) [inline src/mapgen_functions.cpp]: This whole block is terrible and probably copied from somewhere else.
Unless there is a very, very good reason for this, it will need to be all condensed into a neat loop to be tolerable.

And it looks like what you want here is just a separate json.

File: src/mapgen_functions.cpp [hunk 5]
```diff
@@ -5381,3 +5390,335 @@ void madd_field( map *m, int x, int y, field_id t, int density )
...
+        }
+
+    }
+
+    // finally, un-rotate the map
+    m->rotate( rot );
+
+}
+
+void mapgen_railbridge(map *m, oter_id terrain_type, mapgendata dat, int turn, float)
```
Comment (Coolthulhu) [inline src/mapgen_functions.cpp]: Just json those.

File: src/overmap.cpp [hunk 6]
```diff
@@ -381,6 +381,11 @@ bool road_allowed(const oter_id &ter)
     return ter->has_flag( allow_override );
 }
 
+bool railroad_allowed(const oter_id &ter)
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Does the same as `road_allowed` above. Railroads are just different type of road, don't hardcode them as totally separate entities.

File: src/overmap.cpp [hunk 7]
```diff
@@ -577,8 +589,13 @@ bool oter_t::can_connect_to( const int_id<oter_t> &oter ) const
...
     };
-
-    return is_road( id ) ? is_road( oter ) : type_is( *oter->type );
+    // @todo JSONize.
+    const auto is_railroad = []( const int_id<oter_t> &oter ) {
+        return is_ot_type( "railroad",   oter ) ||
+               is_ot_type( "railbridge", oter ) ||
+               is_ot_type( "railhiway",  oter );
+    };
+    return is_road( id ) ? is_road( oter ) : ( is_railroad (id) ? is_railroad ( oter ) : type_is( *oter->type ) );
```
Comment (Coolthulhu) [inline src/overmap.cpp]: This code is terrible, will need that road jsonization instead of this hack.

File: src/overmap.cpp [hunk 8]
```diff
@@ -399,6 +404,13 @@ oter_id overmap::random_house() const
     return one_in( settings.house_basement_chance) ? house : house_base;
 }
 
+oter_id overmap::random_station() const
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Random but returns just one?

File: src/overmap.cpp [hunk 9]
```diff
@@ -588,6 +605,10 @@ bool oter_t::has_connection( om_direction::type dir ) const
     if( id == road_manhole ) {
         return true;
     }
+	static const oter_str_id railroad_manhole("railroad_nesw_manhole");
+	if (id == railroad_manhole) {
+		return true;
+	}
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Hardcoded hack.
Will need jsonization.

File: src/overmap.cpp [hunk 10]
```diff
@@ -1557,6 +1578,11 @@ void overmap::generate(const overmap *north, const overmap *east,
                 roads_out.push_back(city(i.x, 0, 0));
             }
         }
+        for (auto &i : north->railroads_out) {
+            if (i.y == OMAPY - 1) {
+                railroads_out.push_back(city(i.x, 0, 0));
+            }
+        }
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Not strictly necessary here, but it would be much better if this crap was cleaned up before this PR lands.
4 directions, all with their own copypasted handling - ugly, error prone.

File: src/overmap.cpp [hunk 11]
```diff
@@ -1686,6 +1727,8 @@ void overmap::generate(const overmap *north, const overmap *east,
     place_cities();
     place_forest();
 
+	place_stations();
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Tabs, also in other places

File: src/overmap.cpp [hunk 12]
```diff
@@ -1726,6 +1769,45 @@ void overmap::generate(const overmap *north, const overmap *east,
...
+                     is_river(ter(tmp + 1, OMAPY - 1, 0)));
+            viable_railroads.push_back(city(tmp, OMAPY - 1, 0));
+        }
+        if (west == NULL) {
+            do {
+                tmp = rng(10, OMAPY - 11);
+            } while (is_river(ter(0, tmp, 0)) || is_river(ter(0, tmp - 1, 0)) ||
+                     is_river(ter(0, tmp + 1, 0)));
+            viable_railroads.push_back(city(0, tmp, 0));
+        }
```
Comment (Coolthulhu) [inline src/overmap.cpp]: 4 directions, all handled with hardcoded copypasta.
You should handle them all in a neat loop.

File: src/overmap.cpp [hunk 13]
```diff
@@ -3332,6 +3426,72 @@ void overmap::place_cities()
...
+
+            const auto start_dir = om_direction::random();
+            auto cur_dir = start_dir;
+
+            do {
+                build_city_street( cx, cy, size, cur_dir, tmp );
+            } while( ( cur_dir = om_direction::turn_right( cur_dir ) ) != start_dir );
+        }
+    }
+}
```
Comment (Coolthulhu) [inline src/overmap.cpp]: This looks like a giant copy of an existing function - place_cities or something similar. That's a very bad sign.

File: src/overmap.cpp [hunk 14]
```diff
@@ -3719,27 +3879,100 @@ void overmap::connect_closest_points( const std::vector<point> &points, int z, c
...
+            id = node.dir == 1 || node.dir == 3 ? railbridge_ns : railbridge_ew;
+        } else {
+            // @todo Eliminate discrepancy which requires casting and rotation. That is, make 'node' support 'om_direction'.
+            const om_direction::type dir( om_direction::turn_left( static_cast<om_direction::type>( node.dir ) ) );
+            const size_t prev_line( id->type_is( type_id ) ? id->get_line() : 0 );
+
+            id = type_id->get_linear( om_lines::set_segment( prev_line, dir ) );
+        }
+    }
+}
```
Comment (Coolthulhu) [inline src/overmap.cpp]: If this is a copy of build_road_connection or something similar, it will need to be handled by road jsonization. It must not be a copypasted function that duplicates the functionality.

Comment (ZhilkinSerg) [reply src/overmap.cpp]: That is the copy from the March when there were no road jsonization yet. :)

**Comment by Coolthulhu** [comment]
Overall this needs a GIANT cleanup. @codemime has been reworking roads for quite a while, if you wait a while it may be possible to do all of this in a much, much cleaner way.

The huge copy+paste code will just not do. If you insist on doing it now, before road rework, you will need to extract common functions instead of copying existing ones.

**Comment by ZhilkinSerg** [comment]
> Overall this needs a GIANT cleanup. @codemime has been reworking roads for quite a while, if you wait a while it may be possible to do all of this in a much, much cleaner way.

> The huge copy+paste code will just not do. If you insist on doing it now, before road rework, you will need to extract common functions instead of copying existing ones.

Thanks, for the instant review! I know that PR is far from merging. Most of the functions are just a copy of road-mapgen from the beginning of the year. I will surely rework it after I rebase to current master.

File: data/json/overmap_specials.json [hunk 15]
```diff
@@ -1,1770 +1,1867 @@
...
-        "city_sizes" : [4, 14],
-        "occurrences" : [1, 3],
-        "rotate" : false,
-        "spawns" : { "group" : "GROUP_LARGE_STATION", "population":[4, 10], "radius":[1, 2] }
-    },{
-        "type" : "overmap_special",
-        "id" : "warehouse",
-        "overmaps" : [ { "point":[0,0,0], "overmap": "warehouse_north" } ],
-        "connections" : [
-            { "point" : [0
```
Comment (Coolthulhu) [inline data/json/overmap_specials.json]: Looks like you linted the entire file, which made the diff huge.
Linting of this whole file would be useful, but it would need to be a totally separate PR.
For now just manually paste new entries into un-linted version of the file.

Comment (ZhilkinSerg) [reply data/json/overmap_specials.json]: That was a mistake (thought was already linted previously) and I rolled it back.

File: src/mapdata.cpp (hunk 2 above)
Comment (ZhilkinSerg) [reply src/mapdata.cpp]: Some of this is currently being used in the CPP mapgen code. I will try to move most of the mapgen to json and leave only those things in the CPP which are really necessary.

File: src/overmap.cpp (hunk 10 above)
Comment (ZhilkinSerg) [reply src/overmap.cpp]: Now that is just the copy of common roads mapgen - it certainly needs rework and railroads should be based upon reworked version.

**Comment by ZhilkinSerg** [comment]
PR was updated to current master. The code is still ugly and will be refactored. Also there are issues with placing railroad station specials, but code can be compiled.

For the time being railroad tracks are generated very similiar to roads and can replace common roads in cities.

**Comment by ZhilkinSerg** [comment]
Rewoked railroad mapgen functions a bit (bridges are included).

default

**Comment by ZhilkinSerg** [comment]
Overmap mapgen should be much better now - railways are not used as common streets anymore:

default

default

**Comment by Night-Pryanik** [comment]
It maybe just subjective, but railroads on the first overmap pics (straight lines with minimum curves and turns, going right through the forest, not the way round it) look more "natural" than in the later pics.

**Comment by ZhilkinSerg** [comment]
>It maybe just subjective, but railroads on the first overmap pics (straight lines with minimum curves and turns, going right through the forest, not the way round it) look more "natural" than in the later pics.

That's true - screenshots were created using different overmap functions and earlier screenshots were much better. I will definitely make railroads more straight with lesser turns. It will be possible after I add railroad stations to the cities.

**Comment by ZhilkinSerg** [comment]
Railroads should be more straight now. They do not afraid crossing wilderness (plains and forests) and even roads. Only tend to evade swamps, rivers and buildings.

__Overmap version 3__ (`light green` lines - railroads, `light cyan` lines - railroad bridges):

default

__Overmap version 3__ (`light green` lines - railroads, `light cyan` lines - railroad bridges):

default

**Comment by Coolthulhu** [comment]
It seems that railroads override parallel roads. This should not be allowed, we need some way to make it only form crossings with perpendicular ones.

**Comment by ZhilkinSerg** [comment]
> It seems that railroads override parallel roads. This should not be allowed, we need some way to make it only form crossings with perpendicular ones.

Yes, I've also saw this in the latest mapgen. Added this to tasks.

**Comment by ZhilkinSerg** [comment]
Added labels to railroad stations (random city names are used for now).

__Overmap version 4__ (`light green` lines - railroads, `light cyan` lines - railroad bridges, `light green` OO - rail stations stubs):
default

__Overmap version 4__ (`light green` lines - railroads, `light cyan` lines - railroad bridges, `light green` OO - rail stations stubs):
default

**Comment by BorkBorkGoesTheCode** [comment]
Any news?

**Comment by ZhilkinSerg** [comment]
> Any news?

Intersections will come soon.

__Railroad-road intersections__
default

__Overmap version 5__ (`light green` lines - railroads, `light cyan` lines - railroad bridges, `light green` OO - rail stations stubs, `red` - railroad-road intersections):
default

**Comment by Mshock777** [comment]
Shouldn't T-intersection allow for trains to go straight from right to left?
Also Railroad-road intersections generally don't have ties visible.

**Comment by ZhilkinSerg** [comment]
> Shouldn't T-intersection allow for trains to go straight from right to left?

Yes, probably. Will do something later.

> Also Railroad-road intersections generally don't have ties visible.

I've removed ties in the latest commit.

**Comment by DarklordMogrithe** [comment]
How feasible would it be to add railroads to subway mapgen in this PR or a subsequent one? Currently subways are pretty barren, where they would logically have tracks.

**Comment by Zireael07** [comment]
> or a subsequent one?

This, please.

**Comment by ZhilkinSerg** [comment]
> How feasible would it be to add railroads to subway mapgen in this PR or a subsequent one? Currently subways are pretty barren, where they would logically have tracks.

I want to implement all the basic railroad elements first. Once it is done, it will be pretty easy to add more railroad types in subsequent PR.

**Comment by ZhilkinSerg** [comment]
#### Single track railroad station

default

**Comment by tinukedaya** [comment]
Not really sure if it's possible to do, but i'd suggest the the turns are never straight 90 degrees. Making it 45 degrees for few tiles and another 45 degrees after that, would feel much better for Railroads.

**Comment by Night-Pryanik** [comment]
21 files with more than one thousand lines added will be a hard thing to review. Is there a way to split this PR into smaller chunks? E.g. add railroad tiles and make them spawn for step one, then add bridges for step two, stations for step 3 and so on.

**Comment by ZhilkinSerg** [comment]
> 21 files with more than one thousand lines added will be a hard thing to review. Is there a way to split this PR into smaller chunks? E.g. add railroad tiles and make them spawn for step one, then add bridges for step two, stations for step 3 and so on.

I will try to split it.

**Comment by Epictyphlosion** [comment]
I hope this gets worked on again. But this begs a question, how would trains and rail vehicles work?

**Comment by illi-kun** [comment]
What is the purpose of keeping of all of these [WIP] stored in the mergeline for months? Could you please keep only things you're actually working on (to reduce number of active PRs). I have nothing against your work, it's great, but it's a bit confusing to find PRs what actually need a review in the pile of incomplete patches.

**Comment by ZhilkinSerg** [comment]
Close it until #23069 is finished and merged.

**Comment by ZhilkinSerg** [comment]
> What is the purpose of keeping of all of these [WIP] stored in the mergeline for months? Could you please keep only things you're actually working on (to reduce number of active PRs). I have nothing against your work, it's great, but it's a bit confusing to find PRs what actually need a review in the pile of incomplete patches.

I have some reasons to keep them open, but I agree it litters the queue. I will temporary close some PRs with `Stalled` tag.

**Comment by mlangsdorf** [comment]
I'm starting work on solid fuel engine support. Do you have any good numbers for the size, weight, motive power, etc. of steam engines?

**Comment by ZhilkinSerg** [comment]
> I'm starting work on solid fuel engine support. Do you have any good numbers for the size, weight, motive power, etc. of steam engines?

Nope, but I guess we can peep into TTD (http://wiki.openttd.org/Category:Engines).

**Comment by Rivet-the-Zombie** [comment]
What's up with this PR?

**Comment by Rivet-the-Zombie** [comment]
Right on. Good luck with it - it looks really promising!

**Comment by Epictyphlosion** [comment]
Need any help getting this finished?

**Comment by ZhilkinSerg** [comment]
> 
> 
> Need any help getting this finished?

Sure, you can add some loot and monster spawn to railroad station (or add new overmap terrains), add more railroad vehicles or implement railroad points placement/connection algorithm.

**Comment by mount2010** [comment]
Will this still be worked on?

**Comment by ZhilkinSerg** [comment]
> 
> 
> Will this still be worked on?

Yes.

**Comment by JonathanLochridge** [comment]
I saw this was being worked on as of 2019. And this topic came up in the discord so interest seems to be a bit higher for it again.
I expect the ways to help are about the same?
I my skills are mainly with maps and scenarios. Although, I might be able to do a little vehicle work?
But, if you can point me to something I wouldn't mind helping. I have other things I could do if you don't need it.
Or I could try to convince someone to finish cleaning it up if the code is still like how it was mentioned as being earlier and you are done with it?

**Comment by Las-tree** [comment]
Good time of day. I see that you decided not to add railway content, and later abandoned it altogether. Very unfortunate. But is there a chance that you will do it as a separate mod?

**Comment by Las-tree** [comment]
> > Will this still be worked on?
> 
> Yes.

Why you don't work?

**Comment by Zireael07** [comment]
@Las-tree please refrain from rude comments as this one is
